### DISCUSSION TITLE
Host PowerShell modules and scripts in NuGet Gallery?

### DISCUSSION THREAD

**Discussion by o-l-a-v** [PowerShell/PowerShell#24302] [Enhancements]
PowerShell Gallery and its API have various shortcomings and issues, which there does not seem to be resources available for looking into.

## Context

PowerShell Gallery is a fork of NuGet Gallery v2 with modifications/additions to both the web GUI and API:

* <https://learn.microsoft.com/en-us/powershell/gallery/powershellget/supported-repositories?view=powershellget-3.x#the-powershell-gallery>
* <https://github.com/NuGet/NuGetGallery/issues/10071#issuecomment-2227475926>

NuGet v2 has effectively been deprecated by Microsoft since 2015:

<https://learn.microsoft.com/en-us/nuget/api/implementation-guide>

> Note that this guide is focused on the NuGet V3 protocol since the V2 protocol is essentially undocumented and, since 2015, the recommended protocol for NuGet client and server communication is the V3 protocol. For more information read about protocol versioning.

NuGet Gallery has deprecated most features of their v2 API, but the v2 API is still available at `https://www.nuget.org/api/v2`:

* <https://learn.microsoft.com/en-us/nuget/nuget-org/overview-nuget-org#api-endpoint-for-nugetorg>
* 2019-08-07: <https://github.com/NuGet/Announcements/issues/37>
* 2020-12-02: <https://devblogs.microsoft.com/dotnet/custom-v2-odata-queries-will-be-deprecated-march-9-2021/>
* 2021-03-09: <https://github.com/NuGet/Announcements/issues/54>

There have been feature requests to add v3 to PowerShell Gallery:

* <https://github.com/PowerShell/PowerShellGallery/issues/218>

**Would it be time better spent to look at hosting PowerShell modules and scripts in NuGet Gallery instead?** With all the benefits it would bring, like NuGet v3, better API, piggyback uptime and features already in NuGet Gallery.

For a list of related issues I've collected and added to NuGet Gallery as a question for if they thought they could host PowerShell modules and scripts, see:

* <https://github.com/NuGet/NuGetGallery/issues/10071>

With a great reply reply with info on some of the changes that must be looked into for this to happen:

* <https://github.com/NuGet/NuGetGallery/issues/10071#issuecomment-2227475926>

**Comment by rhubarb-geek-nz** [comment]
Although PowerShell modules and dotnet libraries are packaged using the nuget format ( basically a zip file with metadata ) it does not mean they should be hosted together. The layout of a PowerShell module is completely different to that of a dotnet library or framework. They are not meaningfully interchangeable.

They serve very different audiences, NuGet packages for dotnet developers and PowerShell Gallery for PowerShell developers _and end users_.

I see no issue with PowerShell Gallery supporting both NuGet Protocol 3 and NuGet Protocol 2 simultaneously.

I don't think hosting PowerShell modules at nuget.org would benefit either communities.

**Reply by o-l-a-v** [reply]
I'd much rather like PSG to support NuGet v3 and have known issues fixed. But where are the available resources (people, money, will, priority) to do it? There is no public info on roadmap, priorities, ETAs.

They're however looking at adding ACR (Azure Container Registry) support to PSResourceGet, and it has been mentioned that Microsoft is considering to move their official modules to ACR, while keeping PSG a place for community modules and scripts.

ACR is even less suited for PowerShell modules and scripts than NuGet Gallery, AFAIK.

Given the above, I think using NuGet Gallery for PowerShell too would be the least bad option.

**Comment by rhubarb-geek-nz** [comment]
What are the known issues that make the existing PSGallery unusable? As far as I can tell it works.

I don't care if it uses NuGet v2, NuGet v3 or the packages were hosted in a yum or apt repository or plain Azure blobs. What is important is that...

1. when you use the cmdlets provided by PowerShell you can search for, install and update packages
2. there is a UX that is focused on PowerShell users, not dotnet developers

So far nothing is broken from that list.

The only issue I have had with PowerShell packages is hosting them in [github NuGet repository](https://github.com/PowerShell/PowerShell/issues/23834) but I solved that for myself with [rhubarb-geek-nz.NuGetPackage](https://www.powershellgallery.com/packages/rhubarb-geek-nz.NuGetPackage/1.0.2). But being able to use GitHub and lose PSGallery as the central repository would be a loss.

**Reply by o-l-a-v** [reply]
I'm not saying PSG is unusable. If all you care about is getting the latest version of a script or module installed, it works just fine.

If you want to do more advanced things with the API however. There are multiple issues with filtering on versions. I'd say thats core functionality to a package registry, being able to do various things with versions.

Performance:

* Watch paint dry while doing `Get-InstalledPSResource -Path ('{0}\Microsoft\PowerShell\Modules' -f $env:LOCALAPPDATA) | Find-PSResource -Debug` to get the newest version of all your installed modules.
  * The API itself is slow and very inconsistent in performance.
* You can't select only the attributes / properties you want from the API to speed things up ( <https://github.com/PowerShell/PowerShellGallery/issues/267> ).

Uptime. If it's the central place for PowerShell users to get modules and scripts, one should be able to rely on it being up.

---

With this discussion I'm thinking about the future. How will this be in 2 years, 5 years. Right now there are no major issues with PSG, as you say. But by the rate issues are addressed currently, that won't last forever. What would be the best plan moving forward to ensure we have one central place for hosting scripts and modules? Microsoft seems to think Azure Container Registry.

**Reply by rhubarb-geek-nz** [reply]
Lets look at it this way....

1. https://www.powershellgallery.com/ is the URL people go to for a browse

2. PSGallery is the repository people access using the tools

It really does not matter what the backend storage for those artefacts is as long as the cmdlets and UX still function, including Windows PowerShell Desktop 5.1.

I do see problems with migrating packages without the publisher's consent. The model for both nuget.org and PSGallery is that publishers are in control of the packages they publish. I would not recommend automatically migrating packages without the publisher's knowledge or consent and them effectively loosing control over the visibility of (effectively) their property, or worse, a squatter taking their name and publishing a different package.

In terms of "build it and they will come", if people find that hosting their packages at nuget.org is an advantage then let them do so. I have found it is not much help doing so at github.com.

**Reply by o-l-a-v** [reply]
I agree. It should not be a forced migration, but a new place people can opt into start using.
